lib/maths/unittest/CBoostedTreeTest.cc

```diff
@@ -1252,80 +1262,70 @@ void CBoostedTreeTest::testRestoreErrorHandling() {
     auto frame = core::makeMainStorageDataFrame(cols, capacity).first;
 
     std::stringstream errorInBayesianOptimisationState;
-    errorInBayesianOptimisationState
-        << "{\"bayesian_optimization\":"
-           "{\"min_boundary\":{\"dense_vector\":\"-9.191737e-1:-2.041179:-3.506558:1.025:2e-1\"},"
-           "\"max_boundary\":{\"dense_vector\":\"3.685997:2.563991:-1.203973:a:8e-1\"},"
-           "\"error_variances\":\"\",\"kernel_parameters\":{\"dense_vector\":\"1:1:1:1:1:1\"},"
-           "\"min_kernel_coordinate_distance_scales\":{\"dense_vector\":\"1e-3:1e-3:1e-3:1e-3:1e-3\"},"
-           "\"function_mean_values\":{\"d\":\"0\"}},\"best_forest_test_loss\":\"1.797693e308\","
-           "\"current_round\":\"0\",\"dependent_variable\":\"2\",\"eta_growth_rate_per_tree\":\"1.05\","
-           "\"eta\":\"1e-1\",\"feature_bag_fraction\":\"5e-1\",\"feature_sample_probabilities\":\"1:0:0\","
-           "\"gamma\":\"1.298755\",\"lambda\":\"3.988485\",\"maximum_attempts_to_add_tree\":\"3\","
-           "\"maximum_optimisation_rounds_per_hyperparameter\":\"3\",\"maximum_tree_size_fraction\":\"10\","
-           "\"missing_feature_row_masks\":{\"d\":\"3\",\"a\":\"50:0:1:50\",\"a\":\"50:0:1:50\",\"a\":\"50:0:1:50\"},"
-           "\"number_folds\":\"2\",\"number_rounds\":\"15\",\"number_splits_per_feature\":\"40\","
-           "\"number_threads\":\"1\",\"rows_per_feature\":\"50\","
-           "\"testing_row_masks\":{\"d\":\"2\",\"a\":\"50:1:1:5:1:1:5:3:3:3:1:1:1:1:4:1:4:3:6:1:1:2:1:2\","
-           "\"a\":\"50:0:1:5:1:1:5:3:3:3:1:1:1:1:4:1:4:3:6:1:1:2:1:2\"},\"maximum_number_trees\":\"2\","
-           "\"training_row_masks\":{\"d\":\"2\",\"a\":\"50:0:1:5:1:1:5:3:3:3:1:1:1:1:4:1:4:3:6:1:1:2:1:2\","
-           "\"a\":\"50:1:1:5:1:1:5:3:3:3:1:1:1:1:4:1:4:3:6:1:1:2:1:2\"},\"best_forest\":{\"d\":\"0\"},"
-           "\"best_hyperparameters\":{\"hyperparam_lambda\":\"0\",\"hyperparam_gamma\":\"0\","
-           "\"hyperparam_eta\":\"0\",\"hyperparam_eta_growth_rate_per_tree\":\"0\","
-           "\"hyperparam_feature_bag_fraction\":\"0\",\"hyperparam_feature_sample_probabilities\":\"\"},"
-           "\"eta_override\":\"false;0\",\"feature_bag_fraction_override\":\"false;0\",\"gamma_override\":\"false;0\","
-           "\"lambda_override\":\"false;0\",\"maximum_number_trees_override\":\"true;2\",\"loss\":\"mse\"}";
+    readFileToStream("testfiles/error_bayesian_optimisation_state.json",
+                     errorInBayesianOptimisationState);
     errorInBayesianOptimisationState.flush();
 
     bool throwsExceptions{false};
+    std::stringstream buffer;
+    std::streambuf* old = std::cerr.rdbuf(buffer.rdbuf());
```

Comment:
I don't think we should intercept logs like this:

1. It's making the assumption that default logging goes through `std::cerr`, which is a breach of encapsulation.  For example, I was wondering if I should change it to go through `std::clog` so that it's buffered.
2. It's setting the buffer to a local variable which means that if the corresponding `std::cerr.rdbuf(old)` doesn't get run for some reason then the test suite will core dump in the next test.  And when that happens it will be much more work to figure out what went wrong than looking at a nicely formatted list of test failures in Jenkins.

There's an example of how we've previously asserted on which errors are logged in `CPoissonMeanConjugateTest::testSampleMarginalLikelihoodInSupportBounds`.  You could even use the same logging config file in your test as it's also in the `maths` unit tests.

If you don't like this approach then after 7.5 feature freeze I'd be happy for you to open another PR to add a method to `CLogger` to tell it to log to a supplied stream instead of the current logging location and then change the unit tests that use the current approach over to that.  We could have a `catch (...)` in the tests that use that to ensure the logger is reset before the test exits for any reason.  But please use the current approach for this PR and then do that change (if you think it's worthwhile) separately.

Reply:
Thank you for pointing this out, @droberts195. I adjusted the test. Could you please take another look.